modules/epics/src/main/scala/engage/epics/VerifiedEpics.scala

```diff
@@ -0,0 +1,152 @@
+    override val run: F[B]                                         = fa.run.flatMap { a =>
+      fab.run.map { fab =>
+        fab(a)
+      }
+    }
+  }
+
+  case class BindF[F[_]: FlatMap, A, B](
+    fa:   VerifiedEpics[F, A],
+    fafb: F[A] => ChannelTracker[A => F[B]]
+  ) extends VerifiedEpics[F, B] {
+    val fb                                                         = fafb(fa.run)
+    override val systems: Map[TelltaleChannel, Set[RemoteChannel]] = merge(fa.systems, fb.systems)
+    override val run: F[B]                                         = fa.run.flatMap(fb.run)
+  }
+
+  case class IfF[F[_]: FlatMap, A](
+    cond:     F[Boolean],
+    trueVal:  VerifiedEpics[F, A],
+    falseVal: VerifiedEpics[F, A]
+  ) extends VerifiedEpics[F, A] {
+    override val systems: Map[TelltaleChannel, Set[RemoteChannel]] =
+      merge(trueVal.systems, falseVal.systems)
+    override val run: F[A]                                         = cond.flatMap(_.fold(trueVal.run, falseVal.run))
+  }
+
+  case class Get[F[_], A](tt: TelltaleChannel, ch: Channel[F, A]) extends VerifiedEpics[F, A] {
+    override val systems: Map[TelltaleChannel, Set[RemoteChannel]] = Map(tt -> Set(ch))
+    override val run: F[A]                                         = ch.get
+  }
+
+  case class Put[F[_]: FlatMap, A](tt: TelltaleChannel, ch: Channel[F, A], fa: F[A])
+      extends VerifiedEpics[F, Unit] {
+    override val systems: Map[TelltaleChannel, Set[RemoteChannel]] = Map(tt -> Set(ch))
+    override val run: F[Unit]                                      = fa.flatMap(ch.put)
+  }
+
+  case class EventStream[F[_]: Dispatcher: Concurrent, A](tt: TelltaleChannel, ch: Channel[F, A])
+      extends VerifiedEpics[Resource[F, *], Stream[F, StreamEvent[A]]] {
+    override val systems: Map[TelltaleChannel, Set[RemoteChannel]] = Map(tt -> Set(ch))
+    override val run: Resource[F, Stream[F, StreamEvent[A]]]       = ch.eventStream
+  }
+
+  def pureF[F[_]: Applicative, A](v: A): VerifiedEpics[F, A]                            = pure[F[A]](v.pure[F])
+  def unit[F[_]: Applicative]: VerifiedEpics[F, Unit]                                   = pure(Applicative[F].unit)
+  def readChannel[F[_], A](tt: TelltaleChannel, ch: Channel[F, A]): VerifiedEpics[F, A] =
+    Get(tt, ch)
+  def writeChannel[F[_]: FlatMap, A](tt: TelltaleChannel, ch: Channel[F, A])(
+    fa:                                  F[A]
+  ): VerifiedEpics[F, Unit] = Put(tt, ch, fa)
+  def eventStream[F[_]: Dispatcher: Concurrent, A](
+    tt: TelltaleChannel,
+    ch: Channel[F, A]
+  ): VerifiedEpics[Resource[F, *], Stream[F, StreamEvent[A]]] = EventStream(tt, ch)
+  def ifF[F[_]: FlatMap, A](cond: F[Boolean])(trueVal: => VerifiedEpics[F, A])(
+    falseVal:                     => VerifiedEpics[F, A]
+  ): VerifiedEpics[F, A] = IfF(cond, trueVal, falseVal)
+
+  implicit class OpsF[F[_]: Monad, A](v: VerifiedEpics[F, A]) {
```

Comment:
Could `VerifiedEpics` be a `Monad` itself?

Reply:
I couldn't figure out how to do it. My main problem was on how to define `flatMap[A, B](fa: VerifiedEpics[F, A)(f: A => VerifiedEpics[F, B]): VerifiedEpics[F, B]`, because the main idea behind `VerifiedEpics` is that it has the list of channels before evaluating the contained program, but here I need to evaluate the contained program to get the list of channels in the result of `f`.